src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -65,18 +75,100 @@ public class DiscoveryStorage extends AbstractPlatformClientVerticle {
+            Provider<UUID> uuid,
+            FileSystem fs,
+            Path persistencePath,
+            Gson gson,
+            WebClient http,
+            Logger logger) {
         this.uuid = uuid;
+        this.fs = fs;
+        this.persistencePath = persistencePath;
+        this.gson = gson;
+        this.http = http;
         this.logger = logger;
     }
 
     @Override
-    public void start() throws Exception {
-        // TODO persist plugin infos (with empty subtrees) to disk on shutdown, and reinitialize map
-        // here. Then, perform POST on each callback URI to check it's still there and prompt it to
-        // update us with its subtree
+    public void start(Promise<Void> future) throws Exception {
+        for (String s : fs.listDirectoryChildren(persistencePath)) {
+            Path p = persistencePath.resolve(s);
+            try (BufferedReader br = fs.readFile(p)) {
+                PluginInfo pluginInfo = gson.fromJson(br, PluginInfo.class);
+                if (Objects.equals(pluginInfo.getCallback(), NO_CALLBACK)) {
+                    continue;
+                }
+                map.put(UUID.fromString(s), pluginInfo);
+            } catch (IOException ioe) {
+                logger.error(ioe);
+            }
+        }
+
+        map.entrySet()
+                .forEach(
+                        plugin -> {
+                            UUID key = plugin.getKey();
+                            URI uri = plugin.getValue().getCallback();
+                            http.postAbs(uri.toString())
+                                    .timeout(1_000)
+                                    .followRedirects(true)
+                                    .send()
+                                    .onSuccess(
+                                            res -> {
+                                                if (!HttpStatusCodeIdentifier.isSuccessCode(
+                                                        res.statusCode())) {
+                                                    removePlugin(key, uri);
+                                                    return;
+                                                }
+                                                logger.info("Found discovery service {}", uri);
+                                            })
+                                    .onFailure(
+                                            t -> {
+                                                t.printStackTrace();
+                                                removePlugin(key, uri);
+                                            });
+                        });
+        future.complete();
+    }
+
+    private void removePlugin(UUID uuid, URI uri) {
+        map.remove(uuid);
+        try {
+            fs.deleteIfExists(persistencePath.resolve(uuid.toString()));
+        } catch (IOException ioe) {
+            logger.error(ioe);
+        }
+        logger.info("Stale discovery service {} removed", uri);
```

Comment:
Should this log should be called on error tho?

Reply:
I think so - it's still removed from the in-memory map even if the file I/O operation fails.

This is changing a lot in #1031 anyway since there are no more persistence files and in-memory map there, just a proper database connection.

Reply:
oh yeh sure makes sense :D